pkg/resources/pods.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"knative.dev/serving/pkg/apis/serving"
+)
+
+type scopedPodCounter struct {
+	podsLister  corev1listers.PodLister
+	namespace   string
+	serviceName string
+	pods        []*corev1.Pod
+}
+
+// NotRunningPodCounter provides a count of pods currently not in the
+// RUNNING state. The interface exempts users from needing to
+// know how counts are performed.
+type NotRunningPodCounter interface {
+	Get() error
+	TerminatingCount() int
+	PendingCount() int
+}
+
+// NewNotRunningPodsCounter creates a NotRunningPodCounter that counts
+// pods for a namespace/serviceNam. The values returned by
+// TerminatingCount() and PendingCount() will vary over time.
+func NewNotRunningPodsCounter(lister corev1listers.PodLister, namespace, serviceName string) NotRunningPodCounter {
+	return &scopedPodCounter{
+		podsLister:  lister,
+		namespace:   namespace,
+		serviceName: serviceName,
+	}
+}
+
+// Get fetches the list of pods for the given service
+// and caches it in the object
```

Comment:
Looks like we have to call `Get` first before calling `TerminatingCount` and `PendingCount`. Can we hide `Get` method and refactor `TerminatingCount` and `PendingCount` to return (int, error)?

Reply:
tbh, I like it better to do a `Get()` first and then collect counts on the two metrics, compared to how we pull resources for the endpoints. If you compare the code for endpoints and pods, you will notice counting pod metrics has a simpler logic, avoids code duplication, and caches the already fetched resources locally and only once before doing the counting.

The returned error is only from when fetching the list of pods, so if we only need to fetch once to do the counting, I dont think we need to exercise that path twice and return the error twice.

in fact I am tempted to change the code for the endpoints lister too, as it appears to be a cleaner implementation. wdyt?
/cc @vagababov

Reply:
alternatively we can hide `Get` as part of creating the Pod/EndpointCounter objects too, but that overloads the constructor, and arguably goes against best practices.

Reply:
My concern is that it is easily to miss calling `Get`. At least we should document that `Get` should be called before `TerminatingCount` or `PendingCount`, which is absent currently.

Reply:
fair point. updated the comments on these methods to reflect that.